gridfs.go

```diff
@@ -0,0 +1,601 @@
+	grip.DebugWhen(b.opts.Verbose, message.Fields{
+		"type":          "gridfs",
+		"operation":     "reader",
+		"bucket":        b.opts.Name,
+		"bucket_prefix": b.opts.Prefix,
+		"key":           name,
+	})
+
+	grid, err := b.bucket(ctx)
+	if err != nil {
+		return nil, errors.Wrap(err, "resolving bucket")
+	}
+
+	reader, err := grid.OpenDownloadStreamByName(b.normalizeKey(name))
+	if err != nil {
+		if err == gridfs.ErrFileNotFound {
+			err = MakeKeyNotFoundError(err)
+		}
+		return nil, errors.Wrap(err, "opening stream")
+	}
+
+	return reader, nil
+}
+
+func (b *gridfsBucket) Put(ctx context.Context, name string, input io.Reader) error {
+	grip.DebugWhen(b.opts.Verbose, message.Fields{
+		"type":          "gridfs",
+		"dry_run":       b.opts.DryRun,
+		"operation":     "put",
+		"bucket":        b.opts.Name,
+		"bucket_prefix": b.opts.Prefix,
+		"key":           name,
+	})
+
+	grid, err := b.bucket(ctx)
+	if err != nil {
+		return errors.Wrap(err, "resolving bucket")
+	}
+
+	if b.opts.DryRun {
+		return nil
+	}
+
+	if _, err = grid.UploadFromStream(b.normalizeKey(name), input); err != nil {
+		return errors.Wrap(err, "uploading file")
+	}
+
+	return nil
+}
+
+func (b *gridfsBucket) Get(ctx context.Context, name string) (io.ReadCloser, error) {
+	grip.DebugWhen(b.opts.Verbose, message.Fields{
+		"type":          "gridfs",
+		"operation":     "get",
+		"bucket":        b.opts.Name,
+		"bucket_prefix": b.opts.Prefix,
+		"key":           name,
+	})
+
+	return b.Reader(ctx, name)
+}
+
+func (b *gridfsBucket) Upload(ctx context.Context, name, path string) error {
+	grip.DebugWhen(b.opts.Verbose, message.Fields{
+		"type":          "gridfs",
+		"dry_run":       b.opts.DryRun,
+		"operation":     "upload",
+		"bucket":        b.opts.Name,
+		"bucket_prefix": b.opts.Prefix,
+		"key":           name,
+		"path":          path,
+	})
+
+	f, err := os.Open(path)
+	if err != nil {
+		return errors.Wrapf(err, "opening file '%s'", name)
+	}
+	defer f.Close()
+
+	return errors.WithStack(b.Put(ctx, name, f))
+}
+
+func (b *gridfsBucket) Download(ctx context.Context, name, path string) error {
+	grip.DebugWhen(b.opts.Verbose, message.Fields{
+		"type":          "gridfs",
+		"operation":     "download",
+		"bucket":        b.opts.Name,
+		"bucket_prefix": b.opts.Prefix,
+		"key":           name,
+		"path":          path,
+	})
+
+	reader, err := b.Reader(ctx, name)
+	if err != nil {
+		return errors.WithStack(err)
+	}
+
+	if err = os.MkdirAll(filepath.Dir(path), 0700); err != nil {
+		return errors.Wrapf(err, "creating enclosing directory for file '%s'", path)
+	}
+
+	f, err := os.Create(path)
+	if err != nil {
+		return errors.Wrapf(err, "creating file '%s'", path)
+	}
+	_, err = io.Copy(f, reader)
+	if err != nil {
+		_ = f.Close()
+		return errors.Wrap(err, "copying data")
+	}
+
+	return errors.WithStack(f.Close())
+}
+
+func (b *gridfsBucket) Push(ctx context.Context, opts SyncOptions) error {
```

Comment:
I have a sense that, given that `Push` and `Pull` are specifically for the `SyncBucket` interface and are only used for s3.push and s3.pull in the agent (which are unused and may go away due to lack of use), we may want to consider removing `Push`/`Pull` from the bucket interface so we can stop maintaining dead code. Can we file a ticket for this?

Reply:
I think we should keep it for now since it may be useful for local dev / testing (I previously thought we would never use the GridFS impl and, thus, deleted it but here we are LOL). We can always revisit this in https://jira.mongodb.org/browse/EVG-21069.

Reply:
Sounds good, thanks!